### ISSUE TITLE
strict type for insert on duplicate key update

### ISSUE THREAD

**Issue by kir4ik** [juanluispaz/ts-sql-query#85]
Hi,

```ts
import {MySqlConnection} from 'ts-sql-query/connections/MySqlConnection';
import {InsertableValues} from 'ts-sql-query/extras/types';
import {Table} from 'ts-sql-query/Table';
import {StrictOmit} from 'ts-essentials';

class Connection extends MySqlConnection<'Connection'> {
  allowEmptyString = true;
}

class OrderTable extends Table<Connection, 'OrderTable'> {
  id = this.autogeneratedPrimaryKey('id', 'int');
  storeId = this.column('storeId', 'int');
  orderId = this.column('orderId', 'string');
  lineId = this.column('lineId', 'string');
  saProductId = this.column('saProductId', 'int');
  amount = this.column('amount', 'double');
  discount = this.columnWithDefaultValue('discount', 'double');
  quantity = this.column('quantity', 'double');
  unit = this.column('unit', 'string');
  upc = this.optionalColumn('upc', 'string');
  sku = this.optionalColumn('sku', 'string');
  cost = this.optionalColumnWithDefaultValue('cost', 'double');
  stockCount = this.optionalColumn('stockCount', 'double');
  timestamp = this.columnWithDefaultValue('timestamp', 'int');
  isDeleted = this.column('isDeleted', 'boolean');

  constructor() {
    super('sa_order');
  }
}

export const orderTable = new OrderTable();

const ORDER_TABLE_VALUES_FOR_INSERT = orderTable.valuesForInsert();

type RequireAllFields<T> = {
  [K in keyof T & string]-?: T[K];
};

export type OrderNormalizedImportDAOImportParams = {
  orders: RequireAllFields<StrictOmit<InsertableValues<OrderTable>, 'id'>>[];
};

type OrderOnConflictDoUpdateSet = RequireAllFields<
  Pick<typeof ORDER_TABLE_VALUES_FOR_INSERT, keyof OrderNormalizedImportDAOImportParams['orders'][number]>
>;

class OrderNormalizedImportDAO {
  constructor(private getConnection: () => Connection) {}

  async import({orders}: OrderNormalizedImportDAOImportParams): Promise<void> {
    const connection = this.getConnection();

    const orderOnConflictDoUpdateSet: OrderOnConflictDoUpdateSet = {
      amount: ORDER_TABLE_VALUES_FOR_INSERT.amount,
      cost: ORDER_TABLE_VALUES_FOR_INSERT.cost,
      discount: ORDER_TABLE_VALUES_FOR_INSERT.discount,
      isDeleted: ORDER_TABLE_VALUES_FOR_INSERT.isDeleted,
      lineId: ORDER_TABLE_VALUES_FOR_INSERT.lineId,
      orderId: ORDER_TABLE_VALUES_FOR_INSERT.orderId,
      quantity: ORDER_TABLE_VALUES_FOR_INSERT.quantity,
      saProductId: ORDER_TABLE_VALUES_FOR_INSERT.saProductId,
      sku: ORDER_TABLE_VALUES_FOR_INSERT.sku,
      stockCount: ORDER_TABLE_VALUES_FOR_INSERT.stockCount,
      storeId: ORDER_TABLE_VALUES_FOR_INSERT.storeId,
      timestamp: ORDER_TABLE_VALUES_FOR_INSERT.timestamp,
      unit: ORDER_TABLE_VALUES_FOR_INSERT.unit,
      upc: ORDER_TABLE_VALUES_FOR_INSERT.upc
    };

    const orderInsertQuery = connection
      .insertInto(orderTable)
      .values(orders)
      .onConflictDoUpdateSet(orderOnConflictDoUpdateSet);

    await orderInsertQuery.executeInsert();
  }
}
```

The idea is that when refactoring, the TS suggests which fields are missing or, on the contrary, which ones can be removed

Is it possible to add something similar to the onConflictDoUpdateSet method itself so that you can immediately write to `onConflictDoUpdateSet`? Or a public type similar to `OrderOnConflictDoUpdateSet`

**Comment by kir4ik** [comment]
in the method itself, this type is most likely not needed, since there can be composite unique keys in the sql table definition, so the public type would help so that I could build the type I need based on it, instead of my own
```
type OrderOnConflictDoUpdateSet = RequireAllFields<
  Pick<typeof ORDER_TABLE_VALUES_FOR_INSERT, keyof OrderNormalizedImportDAOImportParams['orders'][number]>
>;
```

**Comment by juanluispaz** [comment]
Hi,

What is in the insert and on the update side depends on what you want to achieve; I cannot impose rules on this because it will limit the capabilities of SQL artificially; then, this is something that needs to be managed in an outer layer.

Be aware you have the [valuesForInsert](https://ts-sql-query.readthedocs.io/en/stable/queries/insert/#insert-on-conflict-do-update-upsert) method that allows you to access the values for insert in the update part without using the values directly; that can be useful to create a kind of generic function that enforces what you want.

What I can do is include a utility function in the advanced usage that generates that update set, but I will need to understand more about how it would look like; maybe something like:

- given the insert set
- given a list of properties to exclude (optional) on included
- return an object based on the insert set (excluding the excluded properties) that will contain the same properties to set (if the set property is not in the insert will not be in the update) referencing the corresponding `valuesForInsert`
- of course, the function will return the proper type

Let me know what you think.

**Comment by kir4ik** [comment]
I completely agree with you, it will be enough for me to get access to the OnConflictUpdateSets type

right now I'm inferring a type that only works based on valuesForInsert, but that's not entirely correct since I can't use anything other than valuesForInsert

Thanks

**Comment by juanluispaz** [comment]
On regular libraries, access to the inner types will be easy to provide, but, in ts-sql-query unfortunately, it is not the case; SQL is very complex, and to make it type-safe for you even more; additionally, internally, ts-sq-query uses two different representation to allow TS ingest the types without destroy the performance of the compiler. Besides all of this, the fact to give you access to the inner types put me in a situation where adding support to new SQL constructions is hard; for example, adding upsert support in a typesafe forced me to change inner types in a non-retro-compatible way.

Due to all these reasons, I prefer to provide you with functions that you can use to build your logic upon or simplified versions of types that you can take advantage of in your logic; in the same way, the [utility types](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#utility-types). 

Here I have two options; let me know which one fits the best for you:
Create a utility function in the way I described in the previous message.
Create a utility type like [Updatable row](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#updatable-row) that works in `onConflictDoUpdateSet` using `valuesForInsert`

The most interesting one, and the one I bet on, is probably the second one.

Let me know what you think.

**Comment by kir4ik** [comment]
Fine)
I agree with the second option

Thanks

**Comment by kir4ik** [comment]
Fine)
I agree with the second option

**Comment by juanluispaz** [comment]
Added utility types [`UpdatableOnInsertConflictRow`](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#updatable-row-in-case-of-conflict-on-insert) and [`UpdatableOnInsertConflictValues`](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#updatable-values-in-case-of-conflict-on-insert) to represent updatable values in case of conflict on insert.

Please, share your feedback on this.

Released ts-sql-query 1.49.0 with the changes commented here.

**Comment by kir4ik** [comment]
Great! Thank you very much)
